der/derive/src/sequence.rs

```diff
@@ -50,27 +43,32 @@ impl DeriveSequence {
 
         Self {
             ident: input.ident,
-            lifetime,
+            generics: input.generics.clone(),
             fields,
         }
     }
 
     /// Lower the derived output into a [`TokenStream`].
     pub fn to_tokens(&self) -> TokenStream {
         let ident = &self.ident;
-
-        let lifetime = match self.lifetime {
-            Some(ref lifetime) => quote!(#lifetime),
-            None => default_lifetime(),
+        let mut generics = self.generics.clone();
+
+        // Use the first lifetime parameter as lifetime for Decode/Encode lifetime
+        // if none found, add one.
+        let lifetime: Lifetime = if let Some(lt) = generics.lifetimes().next() {
+            lt.lifetime.clone()
+        } else {
+            let lifetime = default_lifetime();
+            generics.params.insert(
+                0,
+                GenericParam::Lifetime(LifetimeParam::new(lifetime.clone())),
+            );
+            lifetime
         };
```

Comment:
This may not work due to the closure, but what about...

```suggestion
        let lifetime = generics.lifetimes.next().cloned.unwrap_or_else(|| {
            let lt = default_lifetime();
            generics.params.insert(
                0,
                GenericParam::Lifetime(LifetimeParam::new(lt.clone())),
            );
            lt
    });
```

Reply:
Yeah, could go either way... I like it a lot better aside from that mutation